### PR TITLE
[Merged by Bors] - feat(Topology/ShrinkingLemma): add a predicate on refined open sets

### PR DESCRIPTION
**Pull request by yoh-tanimoto**
add a predicate `p : Set X  → Prop` for `PartialRefinent`. All the existing theorems remain valid by taking `fun _ => True` as `p`.

motivation: with this we can require that the refined open set has a compact closure, by taking `fun w => IsCompact (closure w)` as `p`. This can be applied when X is locally compact and T2, and we can obtain a `PartitionOfUnity` for an open cover of a compact set. (This will be done in #12266)

This splits from #12266

### CODE REVIEW

File: Mathlib/Topology/ShrinkingLemma.lean [hunk 1]
```diff
@@ -58,21 +58,23 @@ a maximal element, then show that the maximal element must have `carrier = univ`
   subset_iUnion : s ⊆ ⋃ i, toFun i
   /-- For each `i ∈ carrier`, the original set includes the closure of the refined set. -/
   closure_subset : ∀ {i}, i ∈ carrier → closure (toFun i) ⊆ u i
+  /-- For each `i ∈ carrier`, the refined set satisfies `p`. -/
+  p_refined {i} (hi : i ∈ carrier) : p (toFun i)
```
Comment (YaelDillies) [inline Mathlib/Topology/ShrinkingLemma.lean:62]:
```suggestion
  pred_of_mem {i} (hi : i ∈ carrier) : p (toFun i)
```
maybe?

File: Mathlib/Topology/ShrinkingLemma.lean [hunk 2]
```diff
@@ -144,19 +147,25 @@ def chainSup (c : Set (PartialRefinement u s)) (hc : IsChain (· ≤ ·) c) (ne
       have : v j ≤ v i := (hc.total (hvc _ hxi) (hvc _ hj')).elim (fun h => (hmax j hj' h).ge) id
       simpa only [find_apply_of_mem hc ne (hvc _ hxi) (this.1 <| hiv _ hj')]
   closure_subset hi := (find c ne _).closure_subset ((mem_find_carrier_iff _).2 hi)
+  p_refined {i} hi := by
+    simp only
+    obtain ⟨v, hv⟩ := Set.mem_iUnion.mp hi
+    simp only [mem_iUnion, exists_prop] at hv
+    rw [find_apply_of_mem hc ne hv.1 hv.2]
+    exact v.p_refined hv.2
```
Comment (YaelDillies) [inline Mathlib/Topology/ShrinkingLemma.lean:155]:
```suggestion
  p_refined {i} hi := v.p_refined <| by aesop
```
possibly works?

File: Mathlib/Topology/ShrinkingLemma.lean [hunk 3]
```diff
@@ -144,19 +147,25 @@ def chainSup (c : Set (PartialRefinement u s)) (hc : IsChain (· ≤ ·) c) (ne
...
     (hfin : ∀ x ∈ s, { i | x ∈ u i }.Finite) (hU : s ⊆ ⋃ i, u i) {v} (hv : v ∈ c) :
     v ≤ chainSup c hc ne hfin hU :=
   ⟨fun _ hi => mem_biUnion hv hi, fun _ hi => (find_apply_of_mem hc _ hv hi).symm⟩
 
 /-- If `s` is a closed set, `v` is a partial refinement, and `i` is an index such that
 `i ∉ v.carrier`, then there exists a partial refinement that is strictly greater than `v`. -/
-theorem exists_gt [NormalSpace X] (v : PartialRefinement u s) (hs : IsClosed s) (i : ι)
-    (hi : i ∉ v.carrier) :
-    ∃ v' : PartialRefinement u s, v < v' := by
+theorem exists_gt [NormalSpace X] (v : PartialRefinement u s (fun _ => True)) (hs : IsClosed s)
```
Comment (YaelDillies) [inline Mathlib/Topology/ShrinkingLemma.lean:166]:
```suggestion
theorem exists_gt [NormalSpace X] (v : PartialRefinement u s fun _ => True) (hs : IsClosed s)
```
or even
```suggestion
theorem exists_gt [NormalSpace X] (v : PartialRefinement u s ⊤) (hs : IsClosed s)
```
which I find quite cute

File: Mathlib/Topology/ShrinkingLemma.lean (hunk 2 above)
Comment (yoh-tanimoto) [reply Mathlib/Topology/ShrinkingLemma.lean]: This did not work. Restored the previous code

File: Mathlib/Topology/ShrinkingLemma.lean [hunk 4]
```diff
@@ -144,19 +147,25 @@ def chainSup (c : Set (PartialRefinement u s)) (hc : IsChain (· ≤ ·) c) (ne
...
   ⟨fun _ hi => mem_biUnion hv hi, fun _ hi => (find_apply_of_mem hc _ hv hi).symm⟩
 
 /-- If `s` is a closed set, `v` is a partial refinement, and `i` is an index such that
 `i ∉ v.carrier`, then there exists a partial refinement that is strictly greater than `v`. -/
-theorem exists_gt [NormalSpace X] (v : PartialRefinement u s) (hs : IsClosed s) (i : ι)
-    (hi : i ∉ v.carrier) :
-    ∃ v' : PartialRefinement u s, v < v' := by
+theorem exists_gt [NormalSpace X] (v : PartialRefinement u s ⊤) (hs : IsClosed s)
+    (i : ι) (hi : i ∉ v.carrier) :
+    ∃ v' : PartialRefinement u s (fun _ => True), v < v' := by
```
Comment (YaelDillies) [inline Mathlib/Topology/ShrinkingLemma.lean:168]:
```suggestion
    ∃ v' : PartialRefinement u s ⊤, v < v' := by
```
and below

**Review by YaelDillies — APPROVED** [review]
Thanks!

maintainer delegate

**Pull request closed without merging**